app/services/foreman/provision/ssh.rb

```diff
@@ -1,45 +1,66 @@
-require 'timeout'
+require 'open3'
 
 class Foreman::Provision::Ssh
-  attr_reader :template, :uuid, :results, :address, :username, :options
-
-  def initialize(address, username = "root", options = { })
+  attr_reader :template, :uuid, :address, :username, :options
+
+  # options: {
+  #   template: "/path/to/template",
+  #   uuid:     "uuid",
+  #   key_data: "private-key-content",
+  #   password: "string",
+  # }
+  def initialize(address, username = "root", options = {})
     @username = username
     @address  = address
     @template = options.delete(:template) || raise("must provide a template")
     @uuid     = options.delete(:uuid)     || "#{address}-#{username}"
-    @options  = defaults.merge(options)
+    @options  = options
+    @tmp_key_path = create_tmp_key
 
-    initiate_connection!
+    check_auth_options
+    attempt_connection!
   end
 
   def deploy!
-    logger.debug "about to upload #{template} to remote system at #{remote_script}"
-    scp.upload(template, remote_script)
-    logger.debug "about to execute #{command}"
-    @results = ssh.run(command)
-    log_stdout
-    log_stderr
-    success?
+    scp
+
+    logger.debug "about to execute #{main_command}"
+
+    stdout, stderr, status = Open3.capture3(*ssh_cmd(main_command))
+
+    logger.debug(stdout) if stdout.presence
+    logger.error(stderr) if stderr.presence
+
+    status.success?
   end
 
   private
 
-  def success?
-    return true if results.empty?
-    results.map(&:status).compact == [0]
-  end
+  # Attempt to connect to the remote host and run a simple command
+  def attempt_connection!
+    start_time = Time.now.to_i
+
+    logger.debug "Attempting to connect to #{remote_host}"
+    # TODO: Replace password with ***
+    logger.debug ssh_cmd("pwd")
```

Comment:
This is a must-have

Ftr this is how we dealt with it in sp-rex-ssh https://github.com/theforeman/smart_proxy_remote_execution_ssh/blob/master/lib/smart_proxy_remote_execution_ssh/multiplexed_ssh_connection.rb#L4